### PR TITLE
Strongly-typed shader objects

### PR DESCRIPTION
**Pull request by sectopod**
Having multiple shaders of the same type linked into a program is a classy GL-ism, and we don't need to support it, since no other API does. `ShaderSet` should be superior to the old `Builder`, albeit it totally cuts the output targets specification logic. Considering that you want PSO (#815), this PR would allow easier transition and you can add the targets later on, to PSO itself.

### CODE REVIEW

File: src/core/src/extra/shade.rs [hunk 1]
```diff
@@ -55,7 +55,7 @@ impl<'a> ShaderSource<'a> {
     }
 
     /// Pick one of the stored versions that is the highest supported by the device.
-    pub fn choose(&self, model: ShaderModel) -> Result<&'a [u8], ()> {
```
Comment (csherratt) [inline src/core/src/extra/shade.rs]: I think choose is the correct word.

**Comment by csherratt** [comment]
I like this direction.

Maybe a better approach would be to use a structure rather then an enum.
```rust
struct ShaderSet<R> {
  vertex: Option<VertexShader<R>>,
  geometry: Option<GeometryShader<R>>,
  pixel: Option<PixelShader<R>>,
  targets: &[&str],
  varying: &[&str] 
}
```
This would give you what you want (not allowing linking of two fragment shaders). But without the half dozen combinations that a modern graphics pipe can manage.

**Comment by sectopod** [comment]
I was actually using this `struct` thing in a previous iteration. Aside from the fact it's less ergonomic for the user (have to specify `None` for a lot of stuff each time, and I do expect the user to actually use `ShaderSet`), it's also not as type-safe (i.e. you could attempt forming invalid combinations).

My `enum` approach the only weakness is *potential* implementation bloatware. But firstly, it's not as important as API clarity and user ergonomics, and secondly, there is not really too many combinations in the end (that we need to support). Maybe up to 5 at most.

**Comment by kvark** [comment]
I don't mind either a `struct` or an `enum`, but overall the PR looks to be driving us in the right direction.

**Comment by csherratt** [comment]
Ok. Let's just use an `enum` for now. If it becomes a problem because of a combinational explosion at some point, we can reevaluate it then.

For now this :+1:

**Comment by kvark** [comment]
@sectopod please revert the `choose` change and we'll be happy to proceed

**Comment by sectopod** [comment]
Done. Btw, if we had `struct` full of `Option` members, we'd had to suffer in ergonomics. Not only we'd have to say `None` for the geometry/tessellation shaders every time, but we'd also need to do something with the pixel shader. Formally, it's not required, since we can run with disabled rasterizer. But really, we use it in 99% of cases, thus it would be noisy to always enclose the pixel shader into `Some()`.

**Comment by csherratt** [comment]
That's the allure of the builder patter when you have lots of optional fields.
```rust
Shader::new().pixel(my_pixel_shader).vertex(my_vertex_shader);
```

**Comment by kvark** [comment]
Doesn't look like we have an obviously superior solution here, unfortunately.
`ShaderSet::Simple(my_vertex_shader, my_pixel_shader)` looks nice to me
Let's proceed, see if we can get any closer to PSO and removing of the shader program and batch concepts.
@homu  r+

**Comment by kvark** [comment]
I have a slight dislike for builders since they introduce yet another level of indirection (or layer).

**Pull request merged**